### PR TITLE
publish: evict undeclared extras and pin same-major dest scope

### LINKED ISSUE #2402: extra_pkgs already on undeclaring dests are never evicted; regenerate re-indexes them
## Summary

#2400 closed #2383 for **forward attach**. A CE extra no longer fans into a same-major Plus catalogue whose `extra_pkgs` is `[]`.

The leftover named on the #2383 close (`3c9eb48d`) is still on `origin/devel`: dests that **already** published the extra keep it. Nothing in tagged dest fan-out, Nightly, or `prune_retained` deletes an undeclared dep `.pkg`. A later publish that **does** drop a new canonical and regenerate still re-indexes the stray.

```python
# scripts/catalogue_assembly.py — prune_retained
# Scoped to the canonical package only … a dependency .pkg sitting in the
# same directory (e.g. py311-charset-normalizer-3.4.0.pkg) has no
# independent retention count of its own and is never touched here
```

```python
# scripts/publish_release.py — _drop_assets / publish
# copies only this run's asset_map (canonical + deps that _row_declares_dep).
# existing dest files not in the map are left in place.
# if dest already has the extra AND a new canonical arrives:
#   changed=True → prune_retained (canonical only) → regenerate_catalogue
#   regenerate walks every *.pkg in the dest dir, including the stray extra
```

That is the original Plus harm (#2383): channel repo priority **100** vs Netgate **0**. `add-repo` documents that higher priority wins even at a lower version. A Plus box on `pfblockerng-stable` can still resolve `py311-charset-normalizer` from us.

#2386 (no-op republish) is a subset. Eviction also fails on a **real** drop: new `vX` lands, packagesite is rebuilt, stray extra is still a member.

Same for a CE dest whose row later **drops** `extra_pkgs`: the extra stays on `stable|testing|edge/<ce-varver>/` forever.

Related, not this ticket:

- #2383 — closed completed; forward attach landed in #2400. This is the eviction residual the close already called out. Do not reopen #2383.
- #2386 — no-op republish will not regenerate; eviction would still be required after a regen.
- #2398 — canonical containment heal (copy slower → faster). Opposite axis. Do not fold extras into that heal without an explicit rule.
- #2374 — `not_planned`; orphan 3.3.2 CE copies of the dep are accepted. That ruling does **not** bless leaving the extra on Plus dests.

## Expected

On every tagged / Nightly publish, after attach and **before** `regenerate_catalogue`, for each dest `(channel, varver)` this run targets:

1. Build the desired name set = this run’s canonical + deps that ABI-match **and** `_row_declares_dep` for that dest’s ROUTE row.
2. Unlink any non-catalog `.pkg` whose manifest `name` is not the canonical identity and whose name is not in the desired set (or whose origin is not declared on that row).
3. Then prune canonicals (unchanged) and regenerate.

A dest this run does **not** target is out of scope (do not walk the whole tree unless you add a scheduled audit).

A one-shot Pages delete of the known Plus strays is acceptable to unblock boxes; the publisher still needs the eviction so the next drop cannot put them back into packagesite.

## Reproduction (tip @ `3c9eb48d`)

1. Seed `docs/stable/plus-26.03/py311-charset-normalizer-3.4.0.pkg` (bytes from a CE extra) + a Plus canonical + complete descriptors.
2. Publish a **new** Plus canonical for that varver with `extra_pkgs: []` (the #2383-fixed attach path).
3. **Fail today:** dest still has the extra; regenerated `packagesite` still lists it.
4. `pkg rquery -r pfblockerng-stable '%n' py311-charset-normalizer` on a Plus box subscribed to that dest is non-empty.

## Smoke / tests that would have caught this

**Hermetic — `tests/test_publish_release.py` / Nightly twin (no VM):**

| Seed | Publish | Expect |
|---|---|---|
| `stable/plus-26.03` has Plus canonical `4.0.0.b1` **and** `py311-charset-normalizer-3.4.0.pkg`; Plus row `extra_pkgs=[]` | new Plus canonical `4.0.0.b2` (no extra in asset_map) | extra **gone**; packagesite members do **not** include it; `4.0.0.b2` present |
| same seed | exact republish of `4.0.0.b1` (today a #2386 no-op) | after this ticket: extra gone **or** dest marked dirty so regen runs (pick one; pin it) |
| `edge/ce-2.8` has charset extra; CE row still declares it | new CE canonical | extra **kept** (still declared) |
| `edge/ce-2.8` has charset extra; CE row `extra_pkgs=[]` | new CE canonical | extra **gone** |

Red canary today: first row still has the extra after publish.

**Live / repo-install Plus (after the next real dest write):**

1. Subscribe Plus to `pfblockerng-stable`.
2. `pkg rquery -r pfblockerng-stable '%n' py311-charset-normalizer` — empty if Plus `extra_pkgs` is empty.
3. `pkg rquery -r pfSense '%n' py311-charset-normalizer` — still Netgate’s copy.
4. `pkg install -n -r pfblockerng-stable py311-charset-normalizer` must not win over Netgate.

Do **not** treat `test_first_publish_three_varvers_dependency_only_in_ce` as coverage. It starts from an empty dest and isolates CE/Plus by FreeBSD 15 vs 16.

## Files

- `scripts/publish_release.py` — `publish` / `_drop_assets`
- `scripts/publish_nightly.py` — same dest write
- `scripts/catalogue_assembly.py` — `prune_retained` (canonical-only today) or a sibling undeclared-dep sweep
- `tests/test_publish_release.py`, `tests/test_publish_nightly.py`

### LINKED ISSUE #2403: #2383 acceptance is not on devel; dest tests still isolate extras by ABI
## Summary

#2400 closed #2383 with `_row_declares_dep` on tagged `_build_targets` and Nightly `_route_targets`. Forward attach is the right production change.

The ticket’s **acceptance** is not on `origin/devel` (`3c9eb48d`). The suite can stay green if `_row_declares_dep` is removed from `_build_targets` / `_route_targets` and the helper is left as a unit-only function.

### What landed

`tests/test_issue2383_extra_pkgs_row_scope.py`:

- three helper units (`origin in extra_pkgs`, flavored `py311-` vs `py-`, empty/missing/non-list never declares)
- **import-time mutation** of `ROW_PLUS_03["extra_pkgs"]` (shared list with `ROW_PLUS_07`) so pre-#2383 twin cases still attach `textproc/py311-twin`

`#2400` body: *“Worktree also has integration cases (`test_same_major_dep_not_published_to_row_with_empty_extra_pkgs`, Nightly equivalent); those 47k helper updates are still local-only.”*

`rg test_same_major_dep` on tip: **no matches**.

### Why the existing dest tests stay green without row-scope

```python
# tests/test_publish_release.py
ROW_CE     freebsd_major=15  extra_pkgs=[textproc/py-charset-normalizer]
ROW_PLUS_* freebsd_major=16  extra_pkgs=[]
```

`test_first_publish_three_varvers_dependency_only_in_ce` asserts Plus dests lack `py311-charset-normalizer`. The dep fixture is `FreeBSD:15:*`. ABI alone keeps it off Plus. That was true **before** #2383.

`test_same_dependency_renamed_per_row_with_identical_bytes_publishes` / the different-bytes twin use Plus rows + `textproc/py311-twin`. After row-scope those deps match no dest unless the import-time append runs. Collection order now changes Plus `extra_pkgs` for every later module that imports the same list.

### Matcher leftover of the flavor commit (`bfb240ad`)

```python
# scripts/publish_release.py
def _origin_stem(value):
    last = value.rsplit("/", 1)[-1]
    stripped = _PY_FLAVOR_ORIGIN.sub("", last, count=1)  # ^(py)(?:\d+)?-
    return stripped or last
```

Category is dropped. All of these declare the same extra:

| extra_pkgs | dep origin | `_row_declares_dep` |
|---|---|---|
| `textproc/py-charset-normalizer` | `textproc/py311-charset-normalizer` | True (intended) |
| same | `www/py-charset-normalizer` | **True** |
| same | `security/charset-normalizer` | **True** (no `py-` prefix; last component is already the stem) |

Helper units never pin category.

### Third producer still ABI-only

`scripts/build-repo-portable.py` `build_repo_matrix` still folds `--dep-pkgs` by ABI into every same-major build-role catalog. #2383 named this path. Live Pages now go through `publish_release` / `publish_nightly`. A local `--dep-pkgs` tree still puts a CE extra on every same-major Plus catalog.

Related, not this ticket:

- #2383 — closed completed; do not reopen. Production attach on the two publishers is fixed.
- Eviction of **already-published** Plus extras — sibling ticket (close comment on #2383).
- #2387 — live `--print-route` vs baked `matrix_row` (including `extra_pkgs`). Different axis.

## Expected

1. Land the same-major dest cases #2383 / CodeRabbit asked for. Do **not** keep them local-only.
2. Stop mutating Plus `extra_pkgs` at import. Give the twin tests their own rows that **declare** `textproc/py-twin` (or `textproc/py311-twin` via the flavor rule).
3. `_origin_stem` (or a replacement) must compare `(category, unflavored last)`. `www/py-foo` must not satisfy `textproc/py-foo`. A non-`py-` last component must not satisfy a `py-` extra.
4. Either teach `build_repo_matrix` the same `_row_declares_dep` rule, or document that `--dep-pkgs` is not a dest publisher and fail tests that still ABI-broadcast into a Plus varver with empty `extra_pkgs`.

## Reproduction (tip)

1. Comment out the `_row_declares_dep` conjunct in `_build_targets` / `_route_targets`. Keep the helper and `test_issue2383_*` units.
2. Run `tests/test_publish_release.py::BasicPublishFlowTests::test_first_publish_three_varvers_dependency_only_in_ce` and the 2383 units.
3. **Fail today:** all green. The dest test never shares a major.

## Smoke / tests that would have caught this

**Hermetic — must land on devel (no VM):**

| Setup | Expect |
|---|---|
| CE 2.8 `extra_pkgs=[textproc/py-charset-normalizer]`; Plus 26.03 `extra_pkgs=[]`; **same** `freebsd_major`; dests `["stable","testing","edge"]`; CE-built charset dep | extra exists **only** under `*/ce-2.8/`, not `*/plus-26.03/` |
| same, drop `_row_declares_dep` from `_build_targets` | **RED** |
| Nightly handoff: CE 2.8 leg carries the extra; Plus ROUTE same major, `extra_pkgs=[]` | `nightly/plus-26.03` has no extra; `nightly/ce-2.8` has it |
| same, drop the Nightly `continue` filter | **RED** |
| row `extra_pkgs=[textproc/py-charset-normalizer]`; dep origin `www/py-charset-normalizer` | `_row_declares_dep` is **False** |
| row `extra_pkgs=[textproc/py-charset-normalizer]`; dep origin `textproc/py311-charset-normalizer` | **True** (keep) |
| twin publish with Plus `extra_pkgs=[]` and **no** import mutation | dests have **no** `py311-twin` (or the test rows declare the twin origin explicitly) |
| `build_repo_matrix(..., dep_pkgs=[ce extra])` with CE + Plus same ABI, Plus `extra_pkgs=[]` | Plus catalog **must not** contain the extra (or the call is documented-unsupported and this case is not used) |

Do **not** treat helper units + the 15-vs-16 dest test as #2383 coverage. That is how #2400 closed green without the acceptance table.

**Live (optional):** same Plus `pkg rquery` as the eviction sibling, after a **new** dest that never had the extra. Empty is the attach pin; non-empty on a dest that already had it is the eviction sibling.

## Files

- `scripts/publish_release.py` — `_origin_stem` / `_row_declares_dep` / `_build_targets` error text still says “ABI-matches”
- `scripts/publish_nightly.py` — `_route_targets`
- `scripts/build-repo-portable.py` — `dep_for_abi`
- `tests/test_issue2383_extra_pkgs_row_scope.py` — drop import mutation
- `tests/test_publish_release.py`, `tests/test_publish_nightly.py` — same-major dest cases
- `tests/test_build_repo_portable.py` if the builder path stays a dest producer

### PR DESCRIPTION
**Pull request by BBcan177**
After attach and before `regenerate_catalogue`, tagged and Nightly publish now unlink non-catalog, non-canonical `.pkg` files whose origin the dest ROUTE row does not declare. `prune_retained` stays canonical-only. An eviction marks the dest dirty so an exact republish cannot NOOP over a stale extra.

Same-major dest scope is now on devel, not ABI-isolated 15-vs-16 coverage:
- `_origin_key` compares `(category, unflavored last)` — `www/py-foo` does not satisfy `textproc/py-foo`
- Twin dest tests own `ROW_PLUS_*_TWIN` rows; import-time mutation of shared Plus `extra_pkgs` is gone
- `build_repo_matrix --dep-pkgs` uses the same `_row_declares_origin` rule
- Dropping `_row_declares_dep` from `_build_targets` REDS the new dest case (inspect canary + same-major asset_map spy)

Does not reopen #2383. Does not fold #2398 heal.

Fixes #2402
Fixes #2403
Made-with: Grok

🤖 Generated by Grok and posted on behalf of @BBcan177.

## Summary by CodeRabbit

* **Bug Fixes**
  * Dependency packages are routed only to destinations that explicitly declare them.
  * Improved matching prevents similarly named packages in different categories from being treated as equivalent.
  * Undeclared or stale dependencies are removed during nightly and release publishing while preserving declared packages and catalog files.
  * Dependency mismatch errors now include clearer ABI and package requirement details.

* **Tests**
  * Expanded coverage for dependency routing, cleanup, category isolation, and same-major release scenarios.

### CODE REVIEW

**Comment by BBcan177** [comment]
**Contract conformance** (grok-4.6, effort medium, head `d508417674dc3f7fe74df9189f83612add4bb62d`)

Previous-review lookup: no prior Contract conformance audit comment on this PR. Full-diff review of `d5084176` vs `devel` (`e037329d`).

Verdict: **CLEAN**

## Spec map

Orchestrator spec + #2402 / #2403 acceptance + PR body.

| Claim | Where it lands | Verdict |
| --- | --- | --- |
| Evict undeclared extra `.pkg`s on dests this run targets, before `regenerate_catalogue` | `publish_release._evict_undeclared_deps` + call after `_drop_assets` in `publish_release.publish` and `publish_nightly.publish`; then `prune_retained` + `regenerate_catalogue` only if `changed`. Untargeted dest not walked (`test_untargeted_dest_extra_left_in_place`). | Satisfied |
| Eviction marks dest dirty so exact republish cannot NOOP over a stale extra | Eviction returns `True` → `changed=True` → regen. `test_stale_plus_extra_evicted_on_exact_republish` (tagged + Nightly). | Satisfied |
| Keep declared leftover even when this run did not re-attach it | Helper docstring + skip when `_row_declares_origin`. `test_declared_ce_extra_kept_on_new_canonical` (both publishers). | Satisfied (matches #2402 table row 3; see note below) |
| CE row later drops `extra_pkgs` → extra gone | `test_undeclared_ce_extra_evicted_when_row_drops_extra_pkgs` (both). | Satisfied |
| `prune_retained` stays canonical-only; do not fold #2398 heal; do not reopen #2383 | `catalogue_assembly.py` not in the diff. Backfill still canonical-only (`_iter_canonical_packages`). | Satisfied |
| `_origin_key` is `(category, unflavored last)`; `www/py-foo` does not satisfy `textproc/py-foo` | `publish_release._origin_key` / `build-repo-portable._origin_key`. `test_row_declares_dep_rejects_different_category`. Flavor keep: `test_row_declares_dep_accepts_py_flavor_package_origin`. | Satisfied |
| No import-time Plus `extra_pkgs` mutation; twin tests own their rows | `_declare_twin_on_plus_rows` gone. `ROW_PLUS_03_TWIN` / `ROW_PLUS_07_TWIN` carry `textproc/py-twin`. `test_undeclared_twin_dep_against_empty_plus_extra_pkgs_rejected` uses bare Plus rows. | Satisfied |
| `build_repo_matrix` uses the same declare rule | `dep_for_abi` requires `_row_declares_origin(entry, m.get("origin"))`. `test_dep_pkgs_same_major_plus_empty_extra_pkgs_does_not_receive_ce_extra`. | Satisfied |
| Same-major dest cases on devel | Tagged `SameMajorDestScopeTests` (`ROW_CE` + `ROW_PLUS_SAME_MAJOR`, both FreeBSD 15, dests `stable`/`testing`/`edge`). Nightly `ROW_CE15` + `ROW_PLUS15_03`. | Satisfied |
| Drop `_row_declares_dep` from `_build_targets` is RED | Inspect canary `test_build_targets_match_requires_row_declares_dep` (`"and _row_declares_dep"`). Behavioral + `_drop_assets` asset_map spy in `test_same_major_dep_not_published_to_row_with_empty_extra_pkgs`. Nightly twin: dest case + `test_route_targets_continue_filter_required_for_same_major_plus`. | Satisfied |
| `_build_targets` error text is ABI + declaration, not ABI-only | `"matches no varver targeted … (ABI + extra_pkgs declaration)"`. | Satisfied |

`_origin_stem` is gone from the head tree. `_row_declares_dep` is now a thin wrapper over `_row_declares_origin`; Nightly `_route_targets` calls the shared helper.

## Considered and fine

- **#2402 CodeRabbit item 4** (keep a leftover only if it is both declared *and* in this run's asset_map) is **not** implemented. #2402's own table says a still-declared CE extra is **kept** on a new canonical even when this run does not re-attach it, and the orchestrator spec is “evict **undeclared** extras.” Origin-declaration eviction matches that table and the issue's “or whose origin is not declared” alternative. Not a narrowing.
- **Backfill after evict** cannot put extras back: `backfill_from_slower_channels` copies only `CANONICAL_EMITTED_IDENTITY`.
- **#2402 hermetic table used `stable/plus-26.03`**. Eviction tests drive default `["edge"]`; same `publish()` loop. Same-major dest case covers all three tagged dests.
- **`release_extra_pkgs` stays ABI-only.** That path is older *canonical* releases, not `extra_pkgs` deps. Out of this ticket.

## Findings

None blocking.

### N1 — nitpick — `scripts/publish_release.py:_origin_key` / `scripts/build-repo-portable.py:_origin_key`

#2403 expected §3 also says “a non-`py-` last component must not satisfy a `py-` extra.” `(category, unflavored last)` makes `textproc/charset-normalizer` declare against `textproc/py-charset-normalizer` (both keys `('textproc', 'charset-normalizer')`). The motivating leftover in the ticket table was `security/charset-normalizer`, which is already `False` via category (`test_row_declares_dep_rejects_non_py_last_component_other_category`). Not a #2403 smoke-table row and not the orchestrator origin-key claim. Probe:

```
textproc/py-charset-normalizer     declares=True
textproc/py311-charset-normalizer  declares=True
www/py-charset-normalizer          declares=False
security/charset-normalizer        declares=False
textproc/charset-normalizer        declares=True   # same-category non-py last
```

Suggested fix (only if you want the third sentence independently): require a `py`/`pyNNN-` prefix on both sides before comparing the stripped last, or pin a unit that `textproc/charset-normalizer` is `False`.

### N2 — nitpick — duplicated `_origin_key` / `_row_declares_origin`

Same rule copied into `publish_release.py` and `build-repo-portable.py`. Drift would silently split dest publishers from `--dep-pkgs`. Nightly already imports the publish_release helper. Not a current miss.

## Per-file verdicts

| File | Verdict |
| --- | --- |
| `scripts/publish_release.py` | considered-and-fine — `_origin_key`, `_row_declares_origin`, `_evict_undeclared_deps`, `_build_targets` conjunct, eviction-before-regen |
| `scripts/publish_nightly.py` | considered-and-fine — `_route_targets` `continue` filter + shared evict call |
| `scripts/build-repo-portable.py` | considered-and-fine — `dep_for_abi` uses `_row_declares_origin` |
| `tests/test_publish_release.py` | considered-and-fine — #2402 eviction table, same-major dest + asset_map spy, twin rows, no import mutation |
| `tests/test_publish_nightly.py` | considered-and-fine — Nightly eviction table + same-major dest + continue canary |
| `tests/test_issue2383_extra_pkgs_row_scope.py` | considered-and-fine — category/flavor units + `_build_targets` inspect canary; import mutation removed |
| `tests/test_build_repo_portable.py` | considered-and-fine — existing `--dep-pkgs` rows now declare; new same-major Plus-empty case |

Ledger: no blocking findings.

🤖 Generated by Grok and posted on behalf of @BBcan177.

**Comment by BBcan177** [comment]
**Test honesty** (grok-4.6, effort medium, head `d508417674dc3f7fe74df9189f83612add4bb62d`)

Previous-review lookup: no prior Test honesty audit on this PR. Full-diff pass.

**Verdict: CLEAN**

Mandated mutations were applied one at a time on a `/tmp` worktree of this SHA (host `python3.11` via `PFB_ALLOW_HOST=1`; docker not installed). Baseline before mutation: **22 passed**. After each restore, the tree matched `d5084176` again. No source edits on the PR branch.

### Mandated mutations — all RED

| Mutation | What RED | What stayed green (correct negatives) |
|---|---|---|
| Drop `_evict_undeclared_deps` call sites in `publish()` / Nightly `publish()` | 6 eviction-positive cases: release+nightly `test_stale_plus_extra_evicted_on_{new_canonical,exact_republish}`, `test_undeclared_ce_extra_evicted_when_row_drops_extra_pkgs`. First assertion: `assertFalse(dest/charset exists)` → `AssertionError: True is not false`. | `test_declared_ce_extra_kept_on_new_canonical` (×2), `test_untargeted_dest_extra_left_in_place` |
| Drop `_row_declares_dep` conjunct from `_build_targets` | `test_build_targets_match_requires_row_declares_dep` (`'_row_declares_dep' in source`); `test_same_major_dep_not_published_to_row_with_empty_extra_pkgs` (`CatalogueAssemblyError`: charset missing at `stable/plus-26.03` after attach+evict disagree); `test_undeclared_twin_dep_against_empty_plus_extra_pkgs_rejected` (same identity fault instead of `PublishReleaseError`) | Nightly dest pair (untouched path); helper units |
| Restore import-time `_declare_twin_on_plus_rows()` in `tests/test_issue2383_extra_pkgs_row_scope.py` | `test_undeclared_twin_dep_against_empty_plus_extra_pkgs_rejected` when both files are collected: `PublishReleaseError not raised` (shared `ROW_PLUS_03`/`ROW_PLUS_07` list gained `textproc/py-twin`; dests published `py311-twin`) | Twin `TargetResolutionTests` (own `ROW_PLUS_*_TWIN` lists); Plus charset eviction (twin origin ≠ charset) |

Extra probes (not mandated, same honesty bar): drop `_row_declares_origin` from `build_repo_matrix` `dep_for_abi` REDS `test_dep_pkgs_same_major_plus_empty_extra_pkgs_does_not_receive_ce_extra` (`'py311-charset-normalizer' not in plus_names`). Drop Nightly `_route_targets` continue-filter REDS inspect canary + dest case (identity, same attach+evict shape as tagged).

### Why the dest-case red is still honest

Dropping `_row_declares_dep` does **not** leave the extra sitting on Plus: eviction then unlinks it (row still `extra_pkgs=[]`), and `verify_multi_destination_identity` raises because `source_index` was built from the pre-evict `asset_map`. The spy/`exists`/`packagesite` assertions in `test_same_major_dep_not_published_to_row_with_empty_extra_pkgs` are therefore **unreached** on this mutation. That is still a real red: `_run()` is the production publish path. The `asset_map` spy remains load-bearing if identity is later relaxed — dest-file/`packagesite` checks alone would stay green after eviction-only.

### Findings

None blocking.

**Nitpick 1** — import-mutation canary is collection-coupled.
`test_undeclared_twin_dep_against_empty_plus_extra_pkgs_rejected` (tests/test_publish_release.py:1423) does not import the 2383 module. Isolated `pytest tests/test_publish_release.py::…undeclared_twin…` stayed **green** with the mutation restored. Full-suite / `tests/` collection imports `test_issue2383_*` first and then reds, which is how CI runs and how the old pollution actually fired. A one-liner `import test_issue2383_extra_pkgs_row_scope` + `assert ROW_PLUS_03["extra_pkgs"] == []` would make the pin self-contained.

**Nitpick 2** — no `build-repo-portable` category isolation case.
`www/py-foo` vs `textproc/py-foo` is pinned only on `pr._row_declares_dep` (tests/test_issue2383_extra_pkgs_row_scope.py:48-57). The new `brp` dest test uses matching `textproc/` on both sides, so a stem-only `brp._origin_key` would stay green. Duplicate helper; dest-scope still reds an ABI-only fold.

**Nitpick 3** — Nightly eviction suite has no untargeted-dest keep case. Release `test_untargeted_dest_extra_left_in_place` is the only pin that a whole-tree sweep would fail.

### Per-file verdicts

| File | Verdict |
|---|---|
| `scripts/publish_release.py` | considered-and-fine — eviction call + `_row_declares_dep` conjunct both have executed reds |
| `scripts/publish_nightly.py` | considered-and-fine — eviction call + `_route_targets` continue both have executed reds |
| `scripts/build-repo-portable.py` | considered-and-fine — dest-scope fold reds when `_row_declares_origin` is dropped |
| `tests/test_publish_release.py` | considered-and-fine — new eviction + same-major dest + undeclared-twin cases have failable assertions; TWIN row split is fixture repair, not theater |
| `tests/test_publish_nightly.py` | considered-and-fine — eviction + dest + inspect canary red on the matching drops |
| `tests/test_issue2383_extra_pkgs_row_scope.py` | considered-and-fine — category units would red a stem-only `_origin_key`; inspect canary reds a dropped conjunct; import mutation is gone |
| `tests/test_build_repo_portable.py` | considered-and-fine — new dest case reds an ABI-only fold; existing `--dep-pkgs` tests gained `extra_pkgs` so the new filter still folds (repair, not a dest-scope claim; Plus there is still 16 vs CE 15) |

No `www/` change. No vacuous keep-case: declared-CE keep plants the extra first (`assertTrue` before-state) then republishes without the dep asset. Exact-republish pins `noop is False` **and** extra gone.

🤖 Generated by Grok and posted on behalf of @BBcan177.

**Comment by BBcan177** [comment]
**Correctness + hostile inputs** (grok-4.6, effort medium, head `d508417674dc3f7fe74df9189f83612add4bb62d`)

First pass of this leg (no prior correctness audit on this PR). Full-PR diff reviewed against #2402 / #2403. Attacks executed off-appliance on that SHA (host `python3.11`; Docker not installed).

**Verdict: CLEAN**

No blocking defect. Eviction is dest-scoped, canonical-preserving, and dirty-marks an exact republish so a stray extra cannot sit under a NOOP. Same-major attach is gated by `_row_declares_origin`, not ABI alone. Nightly and tagged share `_evict_undeclared_deps` and behave the same on the named attacks.

## Attacks

| Attack | Result |
|---|---|
| Evict declared extras | **Held.** CE dest that still lists `textproc/py-charset-normalizer` keeps the leftover even when this run does not re-attach it. Flavored `textproc/py311-…` origin matches the unflavored extra. |
| Evict canonical | **Held.** Manifest `name == pfSense-pkg-pfBlockerNG` is skipped. Catalog `packagesite.pkg` / `data.pkg` skipped by name. `meta.conf` is not a `*.pkg`. Filename that only *looks* canonical but whose name is an extra is evicted when undeclared. |
| Dest this run does not target | **Held.** Tagged plus run leaves `edge/ce-2.8` and `nightly/plus-26.03` extras in place. Nightly plus run leaves `nightly/ce-2.8` and `edge/plus-26.03` extras in place. Sweep is `site_root / channel / varver` for this run’s targets only. |
| Category mix-up | **Held** for the cases #2403 named: `www/py-foo` / `www/py311-…` do not satisfy `textproc/py-foo`; `security/charset-normalizer` does not. A www leftover planted beside a declared CE extra is unlinked; the declared extra and its packagesite member stay. Residual same-category non-`py-` stem collision is a nitpick below. |
| Republish NOOP over stray extra | **Held.** Exact republish with a planted extra returns `noop is False`, unlinks the extra, and omits it from `packagesite`. Clean exact republish (no stray) stays NOOP on both publishers. |
| Nightly vs tagged | **Held.** Shared helper. Tagged never walks `nightly/`. Nightly never walks tagged dests. Backfill after tagged evict is canonical-only (`_iter_canonical_packages`) and cannot put an extra back. |

## Findings

### N1 — nitpick — `scripts/publish_release.py` `_origin_key` / `scripts/build-repo-portable.py` same helper

`^(py)(?:\d+)?-` strips the whole `py` / `pyNNN-` prefix, so both `textproc/py-charset-normalizer` and `textproc/charset-normalizer` key to `(textproc, charset-normalizer)`.

```
NOTE  same-category non-py last vs py- extra → True  (ticket said False)
FAIL  same-category non-py leftover evicted (ticket #2403)
      kept because _origin_key strips py- and matches (textproc, charset-normalizer)
```

#2403’s concrete table (www vs textproc, security/other-category) is fixed. This residual only false-keeps on a dest that **already** declares the py extra — Plus `extra_pkgs=[]` still evicts every non-canonical. No production extra uses a non-`py-` last in the same category. Not the leftover Plus-harm.

Suggested fix if you want the ticket’s third sentence literally: normalize `pyNNN-foo` → `py-foo` and leave a last component with no `py-` prefix unchanged.

### N2 — nitpick — `_evict_undeclared_deps` unlinks then reads

A later corrupt `*.pkg` raises `PkgError` after earlier undeclared extras in sort order have already been unlinked. Probe: `aaa-stray.pkg` gone, canonical still present, `mmm-corrupt.pkg` raised. Publish aborts without `regenerate_catalogue`. Fail-closed, dest dirty. Collect then unlink, or unlink only after the scan succeeds.

### N3 — nitpick — Nightly untargeted dest has no suite twin

Tagged has `test_untargeted_dest_extra_left_in_place`. Nightly does not. Executed publish probe confirms Nightly plus does not touch `nightly/ce-2.8` or `edge/plus-26.03`. Coverage hole, not a logic hole.

## Per-file verdicts

| File | Verdict |
|---|---|
| `scripts/publish_release.py` | considered-and-fine (N1/N2 on `_origin_key` / evict scan) |
| `scripts/publish_nightly.py` | considered-and-fine — two-line call into the shared evict; dest is `nightly/<varver>` only |
| `scripts/build-repo-portable.py` | considered-and-fine — `dep_for_abi` now requires `_row_declares_origin`; `release_extra_pkgs` stays ABI-only (older *canonical* releases, not extras) |
| `tests/test_publish_release.py` | considered-and-fine — eviction + same-major + untargeted dest + declared keep |
| `tests/test_publish_nightly.py` | considered-and-fine except N3 |
| `tests/test_issue2383_extra_pkgs_row_scope.py` | considered-and-fine — import-time Plus mutation gone; category units present |
| `tests/test_build_repo_portable.py` | considered-and-fine — `--dep-pkgs` same-major Plus `extra_pkgs=[]` does not receive the CE extra |

## Executed evidence

```
python3.11 -m pytest -q \
  tests/test_issue2383_extra_pkgs_row_scope.py \
  tests/test_publish_release.py::ExtraPkgsEvictionTests \
  tests/test_publish_release.py::SameMajorDestScopeTests \
  tests/test_publish_nightly.py::ExtraPkgsEvictionTests \
  tests/test_publish_nightly.py::SameMajorDestScopeTests \
  tests/test_build_repo_portable.py::test_dep_pkgs_same_major_plus_empty_extra_pkgs_does_not_receive_ce_extra \
  tests/test_build_repo_portable.py::test_cli_dep_pkgs_lands_in_release_and_nightly_same_abi_train \
  tests/test_build_repo_portable.py::test_cli_dep_pkgs_flag_is_repeatable \
  tests/test_build_repo_portable.py::test_dep_pkgs_fold_after_retention_stable_pin_survives
# 23 passed

python3.11 -m pytest -q \
  tests/test_publish_release.py::OutcomeTests::test_exact_republish_is_noop \
  tests/test_publish_nightly.py::NoopTests::test_identical_rerun_is_noop \
  tests/test_publish_nightly.py::RetentionTests::test_retention_evicts_oldest_canonical_dep_survives \
  tests/test_publish_release.py::OutcomeTests::test_retention_evicts_oldest_beyond_keep
# 4 passed
```

Scratch probes (`/tmp/pfb-audit-2429-probe.py`, `/tmp/pfb-audit-2429-probe2.py`) against worktree at `d5084176`:

- Hostile `_origin_key` / `_row_declares_origin` (empty, `None`, non-list, quotes, tabs, traversal, punycode, oversized, `pypy-` / `python-` prefixes): fail-closed except N1.
- Direct `_evict_undeclared_deps`: catalog + canonical kept; declared flavored extra kept; www leftover evicted; undeclared idna evicted; missing-origin evicted; empty `extra_pkgs` evicts leftover; declared leftover kept when not in this run’s asset map.
- Tagged clean republish `noop is True`; stray extra on another varver / nightly dest survives a plus-only tagged run.
- Nightly clean republish `noop is True`; stray extra on targeted dest forces non-NOOP and is unlinked; untargeted nightly + tagged dests untouched; canonical remains.
- Declared CE extra + www leftover: extra kept, www gone, packagesite agrees.

Refute pass: N1 reproduced but downgraded (not Plus-harm; not a named #2403 table row). No remaining claim I could not reproduce as written.

🤖 Generated by Grok and posted on behalf of @BBcan177.

File: tests/test_issue2383_extra_pkgs_row_scope.py [hunk 1]
```diff
@@ -63,6 +45,26 @@ def test_row_declares_dep_accepts_py_flavor_package_origin() -> None:
...
 def test_row_declares_dep_does_not_treat_abi_as_a_declaration() -> None:
     """Vacuity: a row with no extra_pkgs never declares, even for a real origin."""
     assert pr._row_declares_dep({"freebsd_major": "15"}, _dep("textproc/py-charset-normalizer")) is False
+
+
+def test_build_targets_match_requires_row_declares_dep() -> None:
+    """Drop _row_declares_dep from _build_targets must RED the same-major dest case."""
+    source = inspect.getsource(pr._build_targets)
+    assert "_row_declares_dep" in source
+    assert "and _row_declares_dep" in source or "and pr._row_declares_dep" in source
```
[the remark this answers is not in the document]
Comment (BBcan177) [reply tests/test_issue2383_extra_pkgs_row_scope.py:70]: APPLY. Docstrings now state dest-scope behaviour only. Category-mismatch `build_repo_matrix` case added (www/ origin vs textproc/ extra_pkgs → BuildRepoError).

🤖 Generated by Grok and posted on behalf of @BBcan177.
